Randomize timezone in tests across non-deterministic around 1970 and default

azat opened the pull request:
There was some cases when some patches to the datetime code leads to flaky tests, due to the tests itself had been runned using regular timezone (TZ).

But if you will this tests with something "specific" (that is not strictly defined around 1970 year), those tests will fail.

So to catch such issues in the PRs itself, let's randomize session_timezone as well.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Refs: https://github.com/ClickHouse/ClickHouse/issues/42653

azat committed:
Randomize timezone in tests across non-deterministic around 1970 and default

There was some cases when some patches to the datetime code leads to
flaky tests, due to the tests itself had been runned using regular
timezone (TZ).

But if you will this tests with something "specific" (that is not
strictly defined around 1970 year), those tests will fail.

So to catch such issues in the PRs itself, let's randomize
session_timezone as well.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #52184 from azat/ClickHouse:tests/randomize-timezone -->
This is an automated comment for commit d1190b9f0b453d3587632374088dd13b66afaebd with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52184/d1190b9f0b453d3587632374088dd13b66afaebd/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

azat commented:
This is the problem:

```
> GET /?bar=foo/baz&query=select+1+format+Null HTTP/1.1
> User-Agent: curl/8.1.2
```

```
> GET /?bar=foo/baz?query=select+1+format+Null HTTP/1.1
> User-Agent: curl/7.81.0-DEV
```

Now it needs proper encoding for older curl versions

azat committed:
clickhouse-test: add proper escaping for HTTP parameters

The problem is that old versions of cURL (7.81.0 at least) handle
additional parameters incorrectly if in previous parameter was "/":

    $ docker run --rm curlimages/curl:8.1.2 --http1.1 --get -vvv 'http://kernel.org/?bar=foo/baz' --data-urlencode "query=select 1 format Null"; echo
    > GET /?bar=foo/baz&query=select+1+format+Null HTTP/1.1
    > User-Agent: curl/8.1.2

    $ docker run --rm curlimages/curl:7.81.0 --http1.1 --get -vvv 'http://kernel.org/?bar=foo/baz' --data-urlencode "query=select 1 format Null"; echo
    > GET /?bar=foo/baz?query=select+1+format+Null HTTP/1.1
    > User-Agent: curl/7.81.0-DEV

Note, that I thought about making the same for cli, but it is not that
easy, even after getting rid of sh -c and string contantenation, it
still cannot be done for CLICKHOUSE_CLIENT_OPT.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix tests after session_timezone randomization

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on tests/queries/0_stateless/01532_execute_merges_on_single_replica_long.sql:
```diff
@@ -3,6 +3,8 @@
 -- Tag no-parallel: static zk path
 
 SET insert_keeper_fault_injection_probability=0; -- disable fault injection; part ids are non-deterministic in case of insert retries
+-- There is a bug in replication queue, it uses local timezone, will fix it separately
```
https://github.com/ClickHouse/ClickHouse/pull/52325

azat commented:
Interesting, seems that the flakiness of the [`00933_ttl_simple`](https://s3.amazonaws.com/clickhouse-test-reports/52184/aaa3e043e6361e003e86c27956f4b9e8a17b16ad/stateless_tests__msan__[4_6].html) had been fixed by the #52325 as well.

azat commented:
>Interesting, seems that the flakiness of the [00933_ttl_simple](https://s3.amazonaws.com/clickhouse-test-reports/52184/aaa3e043e6361e003e86c27956f4b9e8a17b16ad/stateless_tests__msan__%5B4_6%5D.html) had been fixed by the https://github.com/ClickHouse/ClickHouse/pull/52325 as well.

Still a problem

azat commented:
Test failures unrelated:

>Integration tests (release) [4/4] — fail: 1, passed: 532

- `test_version_update_after_mutation` - failed due to some left overs
